### ISSUE TITLE
Ubuntu 16 not supporting nlohmann/json?

### ISSUE THREAD

**Issue by Tsuyo99** [nlohmann/json#2184]
#### What do you want to achieve?

I would like to install nlohmann-json on ubuntu 16.04 vps server.

My ubuntu info: https://imgur.com/a/uq2kNZG

#### What have you tried?

Tried almost every nlohmann installation command (from what I could find online) to install it from the terminal but nothing wont work.

#### Can you provide a small code example?

https://imgur.com/a/G4sF7d5

#### Which compiler and operating system are you using?

My ubuntu info: https://imgur.com/a/uq2kNZG

I use vpsag.com for a vps server and vnc viewer to view my os.

- Compiler: Terminal?
- Operating system: https://imgur.com/a/uq2kNZG

#### Which version of the library did you use?

- [ ] latest release version 3.7.3
- [ ] other release - please state the version: ___
- [x] the `develop` branch

I'm quite sure I used the dev package command.

P.S: I'm quite new to the whole nlohmann/json thing so I hope I provided you with enough information. I'm here for whatever else you may need!

Thanks in advance,
Thanos

**Comment by nlohmann** [comment]
I am not maintaining that Ubuntu package nor do I know who does and how to update it.

**Comment by FrancoisChabot** [comment]
You can simply do:

`wget https://raw.githubusercontent.com/nlohmann/json/v3.7.3/single_include/nlohmann/json.hpp`

**Comment by nlohmann** [comment]
From https://packages.ubuntu.com/bionic/nlohmann-json-dev I see that @murinicanor could help here.

**Comment by murinicanor** [comment]
> From https://packages.ubuntu.com/bionic/nlohmann-json-dev I see that @murinicanor could help here.

I am the one who created the package for Debian, but that package was superseded by nlohmann-json3-dev (https://tracker.debian.org/pkg/nlohmann-json3). I don't have any knowledge about Ubuntu or how packages get there from the Debian repositories...
@Tsuyo99 maybe you have to enable the universe repository? Maybe that helps: https://askubuntu.com/questions/148638/how-do-i-enable-the-universe-repository

**Comment by Tsuyo99** [comment]
Thank you all for your help! I will check it out tonight and let you know!

**Comment by nlohmann** [comment]
Any news on this?

**Comment by Tsuyo99** [comment]
Sadly nothing worked. I installed Ubuntu 20.04 today and tommorow I'll try the casual installation commands for nlohmann (if it's not already preinstalled or sth).

**Comment by Tsuyo99** [comment]
Not a direct solution to the problem, but I installed Ubuntu 20.04 on my laptop and I successfully installed nlohmann3 so I guess that's something. I'll try to dig deeper onto the ubuntu 16 case tho.

**Comment by lillygeiger** [comment]
There may be a way to accomplish the goal by installing the package that is used by 18.04 bionic, either through pinning or an alternative.  For installation options on past Ubuntu releases, see https://help.ubuntu.com/community/PinningHowto

Please note that there isn't an Ubuntu Backport available at the time of writing to provide support on 16.04 for nlohmann/json. Like Tsuyo99, I'm also experimenting with the potential solutions and will update with what I find.

**Comment by lillygeiger** [comment]
Given the package for 18.04 bionic wasn't tested on 16.04 xenial, it is probably advisable to heed the warnings on the PinningHowto page. Regardless, I was able to successfully install the 18.04 bionic package onto 16.04 xenial by setting up pinning and installing with apt-get install nlohmann-json-dev. Afterward I verified the installation by compiling a C++ program that includes json.hpp and deserializes a JSON input.

Steps:
1. Set the default release (used by pinning) as xenial in /etc/apt/apt.conf (file didn't already exist for me).
Example: `echo 'APT::Default-Release "xenial";' | sudo tee /etc/apt/apt.conf`
2. Add the necessary bionic repository reference in /etc/apt/sources.list
deb http://archive.ubuntu.com/ubuntu bionic main restricted universe multiverse
3. Pin packages in /etc/apt/preferences (file didn't already exist for me).
Package: *
Pin: release n=bionic
Pin-Priority: -10

Package: nlohmann-json-dev
Pin: release n=bionic
Pin-Priority: 1001
4. Suggestion:  check the policy.
sudo apt-cache policy
5. Suggestion:  read man page before first-time usage of pinning (provides pin-priority ranges, and usage - along with explanations - for example pins).
man apt_preferences
6. Refresh apt-get's repositories:
sudo apt-get update
7. Install nlohmann-json-dev:
sudo apt-get install nlohmann-json-dev

Good luck, reader.
